### PR TITLE
Add confirmation to change password

### PR DESCRIPTION
**Pull request by DanielMakinen**
So I made the change password modal that is in style with whats on Figma (except for those toggles that let you see passwords in cleartext. Should i make a new task in Trello to add those in this modal and maybe to ResetPassword modal?)

With this change, that EditableTextWithButtons don't really make sense to me. 
image
Should this just be some kind of a button that opens the modal? I temporarily used linkButton for the functionality.

### CODE REVIEW

**Comment by PatricKangasmaki** [comment]
One issue I found:

image

Now the modal jumps around in size when the error pops, I think it would be a good idea to make the modal size static and the error should just go on two lines if it can't fit. Also, I was wondering if there should be instructions about the password requirements? I think it would be better to have some text like "Password contain atleast 8 characters etc.".

> With this change, that EditableTextWithButtons don't really make sense to me. Should this just be some kind of a button that opens the modal? I temporarily used linkButton for the functionality.  

Never hesitate to contact our designer when in doubt, she's always helpful in things like these 👍

Otherwise this looks great and seems to be working well 👍

**Comment by DanielMakinen** [comment]
> Now the modal jumps around in size when the error pops, I think it would be a good idea to make the modal size static and the error should just go on two lines if it can't fit.

All modals seem to behave this way (at least AddTask and ModifyClient), we might want to make a new card about this if it seems relevant.

> Also, I was wondering if there should be instructions about the password requirements? I think it would be better to have some text like "Password contain atleast 8 characters etc.".

Currently you can see that after you fail to meet the requirements. I agree that it would be good to know before you decide to try some password.

File: frontend/src/components/UserInfoCard.module.scss [hunk 1]
```diff
@@ -35,3 +35,9 @@
   @extend .layout-row;
   min-width: 400px;
 }
+
+.passwordButton {
+  @extend .button-small-filled;
+  justify-content: center;
+  display: flex;
+}
```
Comment (juhoeerola) [inline frontend/src/components/UserInfoCard.module.scss:43]: Please, configure your editor to add newline at the end of the file. That has been the practice on the project, and it also reduces unnecessary noise in git diffs.

Also the indentation is inconsistent in the other css file.

Just formatting the files with prettier should work

File: frontend/src/components/UserInfoCard.tsx [hunk 2]
```diff
@@ -86,27 +86,27 @@ export const UserInfoCard = ({ userInfo }: { userInfo: UserInfo }) => {
   const emailOnOk = async (newValue: string) => {
     dispatch(
       modalsActions.showModal({
-        modalType: ModalTypes.CONFIRM_PASSWORD_MODAL,
+        modalType: ModalTypes.CHANGE_PASSWORD_MODAL,
```
Comment (juhoeerola) [inline frontend/src/components/UserInfoCard.tsx:89]: The change password modal doesn't make sense here.
Previously the modal was just used for asking the user to confirm the change by providing their password.

A separate modal would probably be good for this, since it feels separate from the functionality of changing a password.

File: frontend/src/components/UserInfoCard.tsx [hunk 3]
```diff
@@ -86,27 +86,27 @@ export const UserInfoCard = ({ userInfo }: { userInfo: UserInfo }) => {
...
+          actionData: { id: userInfo.id },
         },
       }),
     );
   };
   const removeOnClick = () => {
     dispatch(
       modalsActions.showModal({
-        modalType: ModalTypes.CONFIRM_PASSWORD_MODAL,
+        modalType: ModalTypes.CHANGE_PASSWORD_MODAL,
```
Comment (juhoeerola) [inline frontend/src/components/UserInfoCard.tsx:109]: Same here as `emailOnOk()`

File: frontend/src/components/UserInfoCard.tsx [hunk 4]
```diff
@@ -86,27 +86,27 @@ export const UserInfoCard = ({ userInfo }: { userInfo: UserInfo }) => {
...
         modalProps: {
           actionData: { id: userInfo.id, email: newValue },
         },
       }),
     );
   };
-  const passwordOnOk = async (newValue: string) => {
+  const passwordOnOk = async () => {
```
Comment (juhoeerola) [inline frontend/src/components/UserInfoCard.tsx:96]: The name `passwordOnOk` doesn't really reflect the current function
maybe `changePassword`, `openChangePasswordModal`, or something like that

Also the `async` is no longer necessary, it was just required by the `EditableTextWithButtons` component
```suggestion
  const changePassword = () => {
```

File: frontend/src/components/modals/ChangePasswordModal.tsx [hunk 5]
```diff
@@ -11,29 +12,35 @@ import { ModalHeader } from './modalparts/ModalHeader';
...
   const formRef = useRef<HTMLFormElement>(null);
 
+  const [errorMessage, setErrorMessage] = useState('');
+  const confirmationError = errorState(useState(''));
+
   const errorHandler = (response?: { data: string; status: number }) => {
     switch (response?.status) {
       case 400: {
-        setErrorMessage(t('Bad request error - is the password correct?'));
+        setErrorMessage(t('Bad request error - is current password correct?'));
```
Comment (juhoeerola) [inline frontend/src/components/modals/ChangePasswordModal.tsx:43]: This didn't have a translation to start with, but it would be good to add while changing the key here

File: frontend/src/components/UserInfoCard.module.scss (hunk 1 above)
Comment (DanielMakinen) [reply frontend/src/components/UserInfoCard.module.scss]: Now I added automatic linebreaks to vscode, and fixed indentation to 2 spaces.

File: frontend/src/components/UserInfoCard.tsx (hunk 2 above)
Comment (DanielMakinen) [reply frontend/src/components/UserInfoCard.tsx]: This is just me being careless. I didn't notice other functions used that modal and just refactored the word.

I need to be more careful about checking how my changes affect other functions in the future.

File: frontend/src/components/UserInfoCard.tsx (hunk 4 above)
Comment (DanielMakinen) [reply frontend/src/components/UserInfoCard.tsx]: You are right, name changed.

File: frontend/src/components/modals/ChangePasswordModal.tsx (hunk 5 above)
Comment (DanielMakinen) [reply frontend/src/components/modals/ChangePasswordModal.tsx]: Translation added.

**Comment by DanielMakinen** [comment]
Thanks for the feedback Juho! Changes were made.

File: frontend/src/components/modals/ChangePasswordModal.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        break;
+      }
+      default:
+        setErrorMessage(t('Internal server error'));
+        break;
+    }
+  };
+
+  const handleConfirm = async () => {
+    if (!formRef.current?.checkValidity()) return;
```
Comment (juhoeerola) [inline frontend/src/components/modals/ChangePasswordModal.tsx:49]: Using the submit functionality on the form might be a better for this. This also allows submitting by pressing <kbd>enter</kbd> like other forms. The form will take care of the validation step, but the default action (posting the form) needs to be prevented.

```suggestion
  const handleConfirm = async (e: SyntheticEvent) => {
    e.stopPropagation();
    e.preventDefault();
```

I guess this was pretty much copied from `ConfirmPasswordModal`, so it would be good to change there as well.

form submit 1/3

File: frontend/src/components/modals/ChangePasswordModal.tsx [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    setIsLoading(false);
+    if (action.rejected.match(res)) {
+      errorHandler(res.payload?.response);
+    } else {
+      closeModal(true);
+    }
+  };
+
+  return (
+    <form ref={formRef} className={classes(css.changePasswordForm)}>
```
Comment (juhoeerola) [inline frontend/src/components/modals/ChangePasswordModal.tsx:72]: form submit 2/3
```suggestion
    <form onSubmit={handleConfirm} className={classes(css.changePasswordForm)}>
```

File: frontend/src/components/modals/ChangePasswordModal.tsx [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+      </ModalContent>
+      <ModalFooter closeModal={closeModal}>
+        <ModalFooterButtonDiv>
+          {isLoading ? (
+            <LoadingSpinner />
+          ) : (
+            <button
+              className="button-large"
+              type="button"
+              onClick={handleConfirm}
```
Comment (juhoeerola) [inline frontend/src/components/modals/ChangePasswordModal.tsx:132]: form submit 3/3
```suggestion
              type="submit"
```

File: frontend/src/components/modals/types.ts [hunk 9]
```diff
@@ -90,6 +91,11 @@ type OwnProps = {
   [ModalTypes.JOIN_PROJECT_MODAL]: { invitation: Invitation };
   [ModalTypes.JOIN_LINK_INVALID_MODAL]: {};
   [ModalTypes.JOIN_LINK_NO_ACCESS_MODAL]: { invitationLink: string };
+  [ModalTypes.CHANGE_PASSWORD_MODAL]: { id: number };
+  [ModalTypes.INFO_MODAL]: { header: string; content: InfoModalContent };
+  [ModalTypes.LEAVE_ROADMAP_MODAL]: {
+    roadmapId: number;
+  };
   [ModalTypes.CONFIRM_PASSWORD_MODAL]:
```
Comment (juhoeerola) [inline frontend/src/components/modals/types.ts:99]: Keeping the unnecessary changes to a minimum helps make sense of the actual change.

In this case, moving the "confirm password modal" stuff back above the new "change password modal" lines would help. (same in `ModalRoot`)

File: frontend/src/components/UserInfoCard.tsx [hunk 10]
```diff
@@ -83,22 +83,22 @@ export const UserInfoCard = ({ userInfo }: { userInfo: UserInfo }) => {
...
+        modalType: ModalTypes.CHANGE_PASSWORD_MODAL,
         modalProps: {
-          actionData: { id: userInfo.id, email: newValue },
+          id: userInfo.id,
         },
       }),
     );
   };
-  const passwordOnOk = async (newValue: string) => {
+  const emailOnOk = async (newValue: string) => {
```
Comment (juhoeerola) [inline frontend/src/components/UserInfoCard.tsx:96]: Here also the reordering should be undone, since `emailOnOk` has not actually changed at all

**Review by juhoeerola — CHANGES_REQUESTED** [review]
There are two things I noticed:
- the form could use `onSubmit` handler on the form, instead of `onClick` on the button
- there is some unnecessary noise from reorderings

Otherwise looks good 👍

File: frontend/src/components/forms/FormField.tsx [hunk 11]
```diff
@@ -69,9 +73,19 @@ function field<T extends FieldType>(
         ref.current.setCustomValidity(error.message);
       }
     }, [ref, error.message, innerRef]);
+    /* eslint-disable */
```
Comment (juhoeerola) [inline frontend/src/components/forms/FormField.tsx:76]: Forgot to ask about this. What does this disable?

File: frontend/src/components/modals/ChangePasswordModal.tsx (hunk 6 above)
Comment (DanielMakinen) [reply frontend/src/components/modals/ChangePasswordModal.tsx]: Changed both ChangePassword and ConfirmPassword modals to use submit.

File: frontend/src/components/forms/FormField.tsx (hunk 11 above)
Comment (DanielMakinen) [reply frontend/src/components/forms/FormField.tsx]: Nothing anymore, I was using it while testing stuff out. I forgot it there. Thanks for pointing that out.

I've started to double check pushes with difftools, hopefully I won't forget these in the future.

**Comment by DanielMakinen** [comment]
Thanks for the feedback Juho!

I implemented all those changes and I think this pull request might be ready now?

**Review by juhoeerola — APPROVED** [review]
Looking good 👍
